consensus/istanbul/core/vrank.go

```diff
@@ -0,0 +1,240 @@
+	view                 istanbul.View
+	committee            istanbul.Validators
+	threshold            time.Duration
+	commitArrivalTimeMap map[common.Address]time.Duration
+}
+
+var (
+	// VRank metrics
+	vrankFirstCommitArrivalTimeGauge           = metrics.NewRegisteredGauge("vrank/first_commit", nil)
+	vrankQuorumCommitArrivalTimeGauge          = metrics.NewRegisteredGauge("vrank/quorum_commit", nil)
+	vrankAvgCommitArrivalTimeWithinQuorumGauge = metrics.NewRegisteredGauge("vrank/avg_commit_within_quorum", nil)
+	vrankLastCommitArrivalTimeGauge            = metrics.NewRegisteredGauge("vrank/last_commit", nil)
+
+	vrankDefaultThreshold = "300ms"
+
+	vrank *Vrank
+)
+
+const (
+	vrankArrivedEarly = iota
+	vrankArrivedLate
+	vrankNotArrived
+)
+
+const (
+	vrankNotArrivedPlaceholder = -1
+)
+
+func NewVrank(view istanbul.View, committee istanbul.Validators) *Vrank {
+	threshold, _ := time.ParseDuration(vrankDefaultThreshold)
+	return &Vrank{
+		startTime:            time.Now(),
+		view:                 view,
+		committee:            committee,
+		threshold:            threshold,
+		commitArrivalTimeMap: make(map[common.Address]time.Duration),
+	}
+}
+
+func (v *Vrank) TimeSinceStart() time.Duration {
+	return time.Now().Sub(v.startTime)
+}
+
+func (v *Vrank) AddCommit(msg *istanbul.Subject, src istanbul.Validator) time.Duration {
+	if v.isTargetCommit(msg, src) {
+		t := v.TimeSinceStart()
+		v.commitArrivalTimeMap[src.Address()] = t
+		return t
+	}
+	return -1
+}
+
+func (v *Vrank) HandleCommitted(blockNum *big.Int) {
+	if v.view.Sequence.Cmp(blockNum) != 0 {
+		return
+	}
+
+	committedTime := v.TimeSinceStart()
+	if v.threshold > committedTime {
+		v.threshold = committedTime
+	}
+
+	vrankQuorumCommitArrivalTimeGauge.Update(int64(committedTime))
+	if len(v.commitArrivalTimeMap) != 0 {
+		sum := int64(0)
+		for _, v := range v.commitArrivalTimeMap {
+			sum += int64(v)
+		}
+		avg := sum / int64(len(v.commitArrivalTimeMap))
+		vrankAvgCommitArrivalTimeWithinQuorumGauge.Update(avg)
+	}
+}
+
+// Log logs accumulated data in a compressed form
+func (v *Vrank) Log() (string, []string) {
+	var (
+		serialized = serialize(v.committee, v.commitArrivalTimeMap)
+		assessed   = assessBatch(serialized, v.threshold)
+		compressed = compress(assessed)
+		bitmap     = hex.EncodeToString(compressed)
+
+		lateCommits = make([]time.Duration, 0)
+		lastCommit  = time.Duration(0)
+	)
+
+	for _, t := range serialized {
+		if assess(t, v.threshold) == vrankArrivedLate {
+			lateCommits = append(lateCommits, t)
+		}
+	}
+
+	// lastCommit = max(lateCommits)
+	for _, t := range lateCommits {
+		if lastCommit < t {
+			lastCommit = t
+		}
+	}
+	if lastCommit != time.Duration(0) {
+		vrankLastCommitArrivalTimeGauge.Update(int64(lastCommit))
+	}
+
+	lateCommitsStrArr := encodeDurationBatch(lateCommits)
+
+	logger.Info("VRank", "seq", v.view.Sequence.Int64(),
+		"round", v.view.Round.Int64(),
+		"bitmap", bitmap,
+		"late", lateCommitsStrArr,
```

Comment:
I see that the bitmap serves as a compressed log of messages, detailing the status of individuals being early, late, or not sent, and it follows the order of the committee. This appears to be intelligent logging for storage save. However, I have some reservations regarding the absence of a convenient method to retrieve the complete set of vrank metrics in the specified sequence. Without such a method, developers might put additional work to make an intactly paired metrics.

Reply:
While your concerns regarding bitmap data retrieval are valid, the bitmap approach is more efficient for data storage. This is beneficial during our temporary use in setting an appropriate VRank threshold on the fluctuating network conditions. 
We will develop a tool for bitmap interpretation, which will address data extraction challenges.

Reply:
We'll make a log parser. However, I'm not sure if the parser needs to be publicly open. The purpose of this PR is solely for our experiments for the reputation system. Could you elaborate on cases where other developers need to use/parse vrank logs?

Reply:
I just thought that a bitmap does sorely not represent the metric fully and intactly itself. Thus I thought some of assistance, matching validators, exact timestamp, etc may need for debugging. 

Rather than making an invert function that returns full metric from a bitmap, I thought that API would be more convenient and robust because the log may be truncated for various reasons (e.g., disk save), or hard to find the exact message.